pkg/oidc/provider/jwks.go

```diff
@@ -0,0 +1,192 @@
+	if err != nil && !errors.IsNotFound(err) {
+		return nil, err
+	}
+
+	if errors.IsNotFound(err) {
+		logrus.Infof("[OIDC provider] creating a new signing key")
+		// generate a default RSA private key
+		privateKey, err := rsa.GenerateKey(rand.Reader, keyBits)
+		if err != nil {
+			return nil, err
+		}
+
+		// Encode private key to PEM
+		privateKeyPEM := pem.EncodeToMemory(&pem.Block{
+			Type:  "RSA PRIVATE KEY",
+			Bytes: x509.MarshalPKCS1PrivateKey(privateKey),
+		})
+
+		// Encode public key to PEM
+		publicKeyDER, err := x509.MarshalPKIXPublicKey(&privateKey.PublicKey)
+		if err != nil {
+			return nil, fmt.Errorf("failed to marshal public key: %w", err)
+		}
+		publicKeyPEM := pem.EncodeToMemory(&pem.Block{
+			Type:  "PUBLIC KEY",
+			Bytes: publicKeyDER,
+		})
+
+		secret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      keySecretName,
+				Namespace: keySecretNamespace,
+			},
+			Data: map[string][]byte{
+				"key.pem": privateKeyPEM,
+				"key.pub": publicKeyPEM,
+			},
+		}
+
+		_, err = secretClient.Create(secret)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &jwksHandler{
+		secretCache:  secretCache,
+		secretClient: secretClient,
+	}, nil
+}
+
+// jwksEndpoint writes the content of the jwks endpoint
+func (h *jwksHandler) jwksEndpoint(w http.ResponseWriter, r *http.Request) {
+	s, err := h.secretCache.Get(keySecretNamespace, keySecretName)
+	if err != nil {
+		logrus.Errorf("[OIDC provider] failed to get secret with public keys %v", err)
+		oidcerror.WriteError(oidcerror.ServerError, "failed to get secret with public keys", http.StatusInternalServerError, w)
+		return
+	}
+	var keys []JWK
+	for name, value := range s.Data {
+		if !strings.HasSuffix(name, ".pub") {
+			continue
+		}
+
+		pubKey, err := getPublicKeyFromSecretData(value)
+		if err != nil {
+			logrus.Errorf("[OIDC provider] failed to extract public key from secret data %v", err)
+			http.Error(w, "failed to extract public key from secret data", http.StatusInternalServerError)
+			return
+		}
+
+		keys = append(keys, JWK{
+			Kty: "RSA",
+			Use: "sig",
+			Kid: strings.TrimSuffix(name, ".pub"),
+			N:   base64.RawURLEncoding.EncodeToString(pubKey.N.Bytes()),
+			E:   base64.RawURLEncoding.EncodeToString(big.NewInt(int64(pubKey.E)).Bytes()),
+		})
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	if err := json.NewEncoder(w).Encode(JWKS{Keys: keys}); err != nil {
+		oidcerror.WriteError(oidcerror.ServerError, "failed to encode JWKS", http.StatusInternalServerError, w)
+	}
+}
+
+// GetSigningKey returns the key used for signing jwt tokens
+func (h *jwksHandler) GetSigningKey() (*rsa.PrivateKey, string, error) {
+	s, err := h.secretCache.Get(keySecretNamespace, keySecretName)
+	if err != nil {
+		return nil, "", err
+	}
+	for name, value := range s.Data {
+		if strings.HasSuffix(name, ".pem") {
+			return getPrivateKeyFromSecretData(name, value)
+		}
+	}
```

Comment:
Hm, I'm not completely convinced about this. I have a strange feeling when I see func  returning more than two values. It looks like the `key.pem` key is fixed when we create the Secret, and also not a guarantee that there won't be other `*.pem` keys in the secret.

Is there any other reason about this?